frostsnapp/native/src/api.rs

```diff
@@ -0,0 +1,196 @@
+pub use crate::coordinator::{
+    PortBytesToReadSender, PortOpenSender, PortReadSender, PortWriteSender,
+};
+pub use crate::FfiCoordinator;
+use anyhow::anyhow;
+use flutter_rust_bridge::{frb, RustOpaque, StreamSink};
+pub use frostsnap_coordinator::PortDesc;
+use lazy_static::lazy_static;
+pub use std::os::fd::RawFd;
+pub use std::sync::{Mutex, RwLock};
+#[allow(unused)]
+use tracing::{event, Level as TLevel};
+
+lazy_static! {
+    static ref EVENT_STREAM: RwLock<Option<StreamSink<CoordinatorEvent>>> = RwLock::default();
+    static ref DEVICE_EVENT_STREAM: RwLock<Option<StreamSink<Vec<DeviceChange>>>> =
+        RwLock::default();
+    static ref PENDING_DEVICE_EVENTS: Mutex<Vec<DeviceChange>> = Default::default();
+}
+
+pub fn init_events(event_stream: StreamSink<CoordinatorEvent>) {
+    let mut v = EVENT_STREAM.write().expect("lock must not be poisoned");
+    *v = Some(event_stream);
+}
+
+pub fn sub_device_events(stream: StreamSink<Vec<DeviceChange>>) {
+    *DEVICE_EVENT_STREAM.write().unwrap() = Some(stream);
+    emit_device_events(vec![]);
+}
+
+pub(crate) fn emit_event(event: CoordinatorEvent) -> anyhow::Result<()> {
+    let stream = EVENT_STREAM.read().expect("lock must not be poisoned");
+
+    let stream = stream.as_ref().expect("init_events must be called first");
+
+    if !stream.add(event) {
+        return Err(anyhow!("failed to emit event"));
+    }
+
+    Ok(())
+}
+
+pub(crate) fn emit_device_events(mut new_events: Vec<DeviceChange>) {
+    let mut events = PENDING_DEVICE_EVENTS.lock().unwrap();
+    events.append(&mut new_events);
+
+    if let Some(stream) = DEVICE_EVENT_STREAM.read().unwrap().as_ref() {
+        let events = std::mem::take(&mut *events);
+        stream.add(events);
+    }
+}
+
+#[frb(mirror(PortDesc))]
+pub struct _PortDesc {
+    pub id: String,
+    pub vid: u16,
+    pub pid: u16,
+}
+
+pub type DeviceId = String;
+
+#[derive(Debug)]
+pub enum CoordinatorEvent {
+    PortOpen { request: PortOpen },
+    PortWrite { request: PortWrite },
+    PortRead { request: PortRead },
+    PortBytesToRead { request: PortBytesToRead },
+}
+
+#[derive(Debug)]
+pub enum DeviceChange {
+    Added { id: DeviceId },
+    Registered { id: DeviceId, label: String },
+    Disconnected { id: DeviceId },
+}
+
+#[derive(Debug)]
+pub struct PortOpen {
+    pub id: String,
+    pub baud_rate: u32,
+    pub ready: RustOpaque<PortOpenSender>,
+}
+
+impl PortOpen {
+    pub fn satisfy(&self, err: Option<String>) {
```

Comment:
I don't quite understand the purpose of this function. Something like, if err is None then the port is open?

Is this the same `.satisfy()` i see being called in the dart code? If so, how does passing bytes into here work?